File: Linked-Lists/Challenge-3.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const processLinkedList = (head) => {
  let current = head;
  while (current !== null) {
    console.log(current);
    current = current.next;
  }
};

const createLinkedList = (values) => {
  if (values.length === 0) {
    return null;
  }
  let head = new Node(values[0]);
  let current = head;
  for (let i = 1; i < values.length; i++) {
    let newNode = new Node(values[i]);
    current.next = newNode;
    current = newNode;
  }
  return head;
};

const values = ['a', 'b', 'c', 'd', 'e'];
const head = createLinkedList(values);

const deleteMiddleNode = (head) => {
  if (head.length === 0) {
    return null;
  }
  let pointerOne = head;
  let pointerTwo = head;
  let maxCount = 0;
  let currentCount = 0;

  while (pointerTwo !== null) {
    maxCount += 1;
    pointerTwo = pointerTwo.next;
  }
  while (pointerOne !== null) {
    let middleNode = Math.ceil(maxCount / 2);
    currentCount += 1;
    if (currentCount + 1 === middleNode) {
      pointerOne.next = pointerOne.next.next;
    }
    pointerOne = pointerOne.next;
  }

  return head;
};

// deleteMiddleNode(head);

const deleteMiddleNode2 = (head) => {
  if (head === null || head.next === null) {
    return null;
  }

  let pointerOne = head;
  let pointerTwo = head;
  let prevPointerOne = null;

  while (pointerTwo !== null && pointerTwo.next !== null) {
    pointerTwo = pointerTwo.next.next;
    prevPointerOne = pointerOne;
    pointerOne = pointerOne.next;
  }
  prevPointerOne.next = pointerOne.next;
  return head;
};

deleteMiddleNode2(head);
processLinkedList(head);

File: Linked-Lists/Challenge-3.py
class Node:
  def __init__(self,val) -> None:
    self.val = val
    self.next = None


def print_linked_list(head):
  current = head
  while current is not None:
    print(current.val)
    current = current.next

def create_linked_list(values):
  head = Node(values[0])
  current = head

  for value in values[1:]:
    new_node = Node(value)
    current.next = new_node
    current = new_node
  return head

values = ['A','B','C','D','E']
head = create_linked_list(values)

# O(N) Time Complexity and 0(1) Space Complexity

def delete_middle_node(head):
  pointer_one = head
  pointer_two = head
  previous_node = None


  while pointer_two is not None and pointer_two.next is not None:
    previous_node = pointer_one
    pointer_one = pointer_one.next
    pointer_two = pointer_two.next.next

  if previous_node:
    previous_node.next = pointer_one.next
  else:
    head = pointer_one.next

  return head

newhead = delete_middle_node(head)
print_linked_list(newhead)

